diffkemp/building/cc_wrapper_template.py

```diff
@@ -0,0 +1,213 @@
+        # Compile/link mode with object files detected
+        # Drop object files and revert to normal compile/link mode
+        clang = old_clang
+        new_clang_argv = []
+        for arg in clang_argv:
+            if arg.endswith('.ll'):
+                continue
+            new_clang_argv.append(arg)
+        clang_argv = new_clang_argv
+
+    # Do not continue if output is not .ll or .llw
+    # Note: this means that this is neither compilation nor linking
+    if (output_file is not None and not output_file.endswith('.ll') and
+            not output_file.endswith('.llw')):
+        return 0
+
+    # Do not run clang on conftest files
+    if output_file in ['conftest.ll', 'conftest.llw'] or 'conftest.c' in argv:
+        return 0
+
+    # Record file in database
+    if output_file is not None:
+        prefix = 'o:' if clang != llvm_link else 'f:'
+        db.append(prefix + os.path.join(os.getcwd(), output_file))
+    elif not linking:
+        # Compiling to default output file
+        for arg in clang_argv:
+            if not arg.endswith('.c'):
+                continue
+            db.append('o:' + os.path.join(os.getcwd(),
+                                          arg.rsplit('.', 1)[0] + '.ll'))
+
+    # Analyze and modify parameters for clang (phase 2)
+    clang_argv[0] = clang
+    if clang != llvm_link:
+        for arg in append:
+            clang_argv.append(arg)
+        # TODO: allow compiling into binary IR
+        clang_argv.append('-S')
+        clang_argv.append('-emit-llvm')
+    else:
+        # Keep only arguments with input files (and llvm-link itself)
+        clang_argv = [arg for arg in clang_argv if arg == clang or
+                      arg.endswith('.ll') or arg.endswith('.llw') or
+                      arg == '-o']
+        # Remove non-existent files
+        # Note: these might have been e.g. generated from assembly
+        new_clang_argv = [clang_argv[0]]
+        output_file_flag = False
+        for arg in clang_argv[1:]:
+            if output_file_flag or os.path.exists(arg) or arg == '-o':
+                new_clang_argv.append(arg)
+                if output_file_flag:
+                    output_file_flag = False
+            if arg == '-o':
+                output_file_flag = True
+        clang_argv = new_clang_argv
+
+    # Run clang
+    try:
+        if debug:
+            print("Wrapper calling: " + ' '.join(clang_argv))
+        check_call(clang, clang_argv)
+    except CalledProcessError:
+        print("cc_wrapper: warning: clang failed")
+        print(clang_argv)
+        # clang error is non-fatal
+        return 0
+
+    # Update database
+    with open(db_filename, 'a') as db_file:
+        # TODO: RPython-compatible file locking
+        for entry in db:
+            if not os.path.exists(entry.split(':')[1]):
+                continue
+            db_file.write(entry + '\n')
+
+    # Disassemble output when running llvm-link
```

Comment:
Does this assume that `llvm-link` always produces assembled (`.bc`) files? Is that really true?

Reply:
Yes, unless the `-S` option is used, I could use that instead of disassembling with `llvm-dis`.